assets/js/amp-block-validation.js

```diff
@@ -0,0 +1,166 @@
+/**
+ * Validates blocks for AMP compatibility.
+ *
+ * This uses the REST API response from saving a page to find validation errors.
+ * If one exists for a block, it display it inline with a Notice component.
+ */
+
+/* exported ampBlockValidation */
+/* global wp, _ */
+var ampBlockValidation = ( function() {
+	'use strict';
+
+	var module = {
+
+		/**
+		 * Holds data.
+		 */
+		data: {
+			i18n: {
+				invalidAmpContentNotice: ''
+			}
+		},
+
+		/**
+		 * Boot module.
+		 *
+		 * @param {Object} data - Module data.
+		 * @return {void}
+		 */
+		boot: function boot( data ) {
+			module.data = data;
+
+			wp.hooks.addFilter(
+				'blocks.BlockEdit',
+				'amp/add-notice',
+				module.conditionallyAddNotice
+			);
+		},
+
+		/**
+		 * Wraps the edit() method of a block, and conditionally adds a Notice.
+		 *
+		 * @param {Function} BlockEdit - The original edit() method of the block.
+		 * @returns {Function} The edit() method, conditionally wrapped in a notice for AMP validation error(s).
+		 */
+		conditionallyAddNotice: function conditionallyAddNotice( BlockEdit ) {
+			return function( props ) {
+				var errors, result;
+
+				errors = module.getBlockValidationErrors( props );
+				result = [
+					wp.element.createElement( BlockEdit, _.extend({}, props, { key: 'amp-original-edit' }) )
+				];
+
+				if ( 0 < errors.length ) {
+					result.unshift(
+
+						// @todo Add PanelBody with validation error details.
+						wp.element.createElement(
+							wp.components.Notice,
+							{
+								status: 'warning',
+								isDismissible: false,
+								key: 'amp-validation-notice'
+							},
+							module.data.i18n.invalidAmpContentNotice + ' ' + _.unique( _.pluck( errors, 'code' ) ).join( ', ' )
+						)
+					);
+				}
+				return result;
+			};
+		},
+
+		/**
+		 * Gets the block types with errors.
+		 *
+		 * Iterates through the 'amp_validation_errors' from the REST API response.
+		 * This returns an object, with block types as the keys, and error arrays as the values.
+		 * The block's overridden edit() method can then get the errors for its block type.
+		 * This finds the validation errors for a specific block, based on its attributes or content.
+		 * @todo: keep refining how this finds if the errors match.
```

Comment:
I think I know what we need to do here. We need to grab the validation errors, group them by `postId` and `blockIndex`, then align with `getBlockOrder` upon update of the `restValidationErrorsField`. I'll push up a bit more on this tomorrow.